File: csharp/UniQ.cs
/**
* A Lock-free Queue
*/ 

using System;
using System.Threading;
using System.Diagnostics;

namespace UniQ
{
  public class Queue
  {
    protected int[] data;
    protected int head, tail;
    protected int mask;

    public int size { get; private set; }
    public int count { get { return tail - head; } }

    public Queue(int size = 8)
    {
      this.size = size;
      mask = this.size - 1;
      tail = head = 0;
      Array.Resize(ref data, this.size);
    }
  
    public virtual void push(int item)
    {
      Debug.Assert(item != 0, "Queue.push(item): item can't be null");
      int t;
      do {
        t = tail;
        while (t - head == size){ };// { Thread.Sleep(0); };
      } while ( (data[t & mask]!=0) || (Interlocked.CompareExchange(ref tail, t+1, t) != t) );
      data[t & mask] = item;
    }

    public virtual int pop()
    {
      int h;
      do {
        h = head;
        while (h == tail){ };// { Thread.Sleep(0); };
      } while ( (data[h & mask]==0) || (Interlocked.CompareExchange(ref head, h+1, h) != h) );

      h &= mask;
      int r = data[h];
      data[h] = 0;
      return r;
    }
  };
};

// Part of uniQ library released under GNU 3.0 

File: js/uniq.js
// This is a early attemp to migrate uniQ lib to JavaScript. 
// See README.md for details - Released under GNU 3.0

const sleep = () => new Promise(resolve => setImmediate(resolve))
  , assert = require('assert')

function Queue(buffer = 64) {
  // in JS, we use a single buffer to hold data with input & output at the end
  // the parameter buffer can be the size fora a new buffer or a SharedArrayBuffer
  if (typeof buffer == 'number')
    buffer = new SharedArrayBuffer(Int32Array.BYTES_PER_ELEMENT * (buffer + 2))

  var data = new Int32Array(buffer)
    , length = data.length - 2
    , mask = length - 1
    , input = length
    , output = length + 1

  assert(length && !(length & (length - 1)), "Queue size must be a positive power of 2.");

  async function push(item) {
    assert(typeof item == 'number' && item !== 0)
    return new Promise(async resolve => {
      let i = 0;
      do {
        i = data[input];
        while (i - data[output] > mask) // if full, wait for space
          await sleep() // Atomics.wait(input, 0, i) 
      } while (data[i & mask] || (Atomics.compareExchange(data, input, i, i + 1) !== i))

      data[i & mask] = item
      return resolve(i)
    })
  }

  async function pop() {
    return new Promise(async resolve => {
      let o;
      do {
        o = data[output]
        while (o == data[input]) // if empty, wait for items
          await sleep() // Atomics.wait(output, 0, o)
      } while (!data[o & mask] || Atomics.compareExchange(data, output, o, o + 1) !== o)

      let r = data[o &= mask]
      data[o &= mask] = 0
      return resolve(r);
    })
  }
  return { buffer, push, pop }
}
module.exports = { Queue }